pkg/recommendation/device.go

```diff
@@ -0,0 +1,250 @@
+
+	if request.Spec.PoolCapacity.IsZero() {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got zero pool capacity")
+	}
+	if data.BlockDeviceList == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil block device list")
+	}
+
+	err := request.Spec.DataConfig.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create capacity recommendation request")
+	}
+
+	cspcrr := cStorPoolClusterRecommendationRequest{
+		Request: types.CStorPoolClusterRecommendationRequest{
+			Spec: types.CStorPoolClusterRecommendationRequestSpec{
+				PoolCapacity: request.Spec.PoolCapacity,
+				DataConfig:   request.Spec.DataConfig,
+			},
+		},
+		Data: *data,
+	}
+
+	return &cspcrr, nil
+}
+
+// GetRecommendation returns recommended block devices for all nodes and device types.
+func (r *cStorPoolClusterRecommendationRequest) GetRecommendation() map[string]types.CStorPoolClusterRecommendation {
+
+	cStorPoolClusterRecommendation := make(map[string]types.CStorPoolClusterRecommendation)
+
+	if len(r.Data.BlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	if err := r.Request.Spec.DataConfig.Validate(); err != nil {
+		return cStorPoolClusterRecommendation
+	}
+
+	availableBlockDeviceList := unstructured.UnstructuredList{}
+	availableBlockDeviceList.Object = r.Data.BlockDeviceList.Object
+	for _, bd := range r.Data.BlockDeviceList.Items {
+		isEligible, err := bdutil.IsEligibleForCStorPool(bd)
+		if err == nil && isEligible {
+			availableBlockDeviceList.Items = append(availableBlockDeviceList.Items, bd)
+		}
+	}
+	if len(availableBlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	deviceTypeNodeBlockDeviceMap := bdutil.GetTopologyMapGroupByDeviceTypeAndBlockSize(availableBlockDeviceList)
+	if len(deviceTypeNodeBlockDeviceMap) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	for kind, nodeBlockDeviceListMap := range deviceTypeNodeBlockDeviceMap {
+
+		nodeCapacityBlockDeviceMap := nodeCapacityBlockDevices{}
+
+		for nodeName, blockDeviceList := range nodeBlockDeviceListMap {
+
+			// capacityBlockDevicesMap contains block device capacity and all block devices
+			// of that capacity in one node.
+			capacityBlockDevicesMap := nodeCapacityBlockDeviceMap.getOrDefault(nodeName)
+
+			for _, blockDevice := range blockDeviceList {
+				capacity, found := blockDevice.Capacity.AsInt64()
+				if !found {
+					// TODO handle this case
+					continue
+				}
+
+				blockDevices := capacityBlockDevicesMap.getOrDefault(capacity)
+				capacityBlockDevicesMap.update(capacity, append(blockDevices, blockDevice))
+
+			}
+
+			// update nodeCapacityBlockDeviceMap for each node.
+			nodeCapacityBlockDeviceMap.update(nodeName, capacityBlockDevicesMap)
+		}
+
+		cStorPoolClusterRecommendationValue := nodeCapacityBlockDeviceMap.getDeviceRecommendtion(r.Request.Spec.PoolCapacity, r.Request.Spec.DataConfig)
```

Comment:
should we have a check for an empty pool instance?
ie - If we can not create ccc for a given raid group config then we should not add an empty response.
A simple example I have 1 block device connected with node but I want mirror pool so I can not create ccc for that.
From this code, I can see we are adding empty config(empty instance)
We can add a check If instance length is 0 then do not add that in the recommendation.

Reply:
typo - `getDeviceRecommendtion`